src/main/webapp/js/index/index.js

```diff
@@ -0,0 +1,108 @@
+        })
+        .catch((error) => {
+            errorEl.classList.remove("success-banner");
+            errorEl.classList.remove("hidden");
+            errorEl.classList.add("error-banner");
+            errorEl.innerText = error;
+        });
+}
+
+/* ==========================================================================
+   USER LOCATION AND ADDRESS
+   ========================================================================== */
+function getLocation() {
+    if (navigator.geolocation)
+        navigator.geolocation.getCurrentPosition(geoLocEnabled, geoLocFallback);
+    else
+        geoLocFallback();
+}
+
+// Geolocation is supported and enabled
+function geoLocEnabled(position) {
+    let locationEl = document.getElementById("location-container");
+    let pos = {
+        lat: position.coords.latitude,
+        lng: position.coords.longitude,
+    };
+    localStorage.setItem("lat", pos.lat);
+    localStorage.setItem("lng", pos.lng);
+    convertLocation(pos).then((address) => { locationEl.innerText = address; });
+}
+
+// Use inaccurate IP-based geolocation instead
+function geoLocFallback() {
+    let locationEl = document.getElementById("location-container");
+    $.ajax({
+        type: "POST",
+        url: 'https://www.googleapis.com/geolocation/v1/geolocate?key=' + 'AIzaSyBL_9GfCUu7DGDvHdtlM8CaAywE2bVFVJc', // TODO: use safe API key storage!!!
+        data: { considerIp: 'true' },
+        success: function(response) {
+            let pos = {
+                lat: response.location.lat,
+                lng: response.location.lng,
+            };
+            localStorage.setItem("lat", pos.lat);
+            localStorage.setItem("lng", pos.lng);
+            convertLocation(pos).then((address) => { locationEl.innerText = address; });
+        },
+        error: function(xhr) {
+            if (xhr.status == 404)
+                console.log("No results");
+            if (xhr.status == 403)
+                console.log("Usage limits exceeded");
+            if (xhr.status == 400)
+                console.log("API key is invalid or JSON parsing error");
+            geoLocHardcoded(); // DEBUG
+        }
+    });
+}
+
+// TODO: remove this entirely; fallback should be to leave form blank instead
+function geoLocHardcoded() {
+    let locationEl = document.getElementById("location-container");
+    let pos = { lat: 40.730610, lng: -73.935242 };
+    localStorage.setItem("lat", pos.lat);
+    localStorage.setItem("lng", pos.lng);
+    convertLocation(pos).then((address) => {
+        locationEl.innerText = address;
+    });
+}
+
+// Convert lat/lng to human readable address
+function convertLocation(location) {
+    let lat = location.lat;
+    let long = location.lng;
+    return fetch(`/convert?lat=${lat}&lng=${long}`)
+        .then((response) => response.json())
+        .then((response) => { return response.results[0].formatted_address; })
+        .catch((error) => console.log(error));
```

Comment:
this is fine for now I suppose, but we should try to surface this error to the user

Reply:
sure, I will change this in the next PR with merging Emma's stuff

Reply:
Can you add a TODO in the code?